// @stryker-mutator/html-reporter/src/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var util_1 = require("@stryker-mutator/util");
var path = require("path");
var mkdirp = require("mkdirp");
var rimraf = require("rimraf");
function copyFile(fromFilename, toFilename) {
    return new Promise(function (resolve, reject) {
        var readStream = util_1.fsAsPromised.createReadStream(fromFilename);
        var writeStream = util_1.fsAsPromised.createWriteStream(toFilename);
        readStream.on('error', reject);
        writeStream.on('error', reject);
        readStream.pipe(writeStream);
        readStream.on('end', resolve);
    });
}
exports.copyFile = copyFile;
function deleteDir(dirToDelete) {
    return new Promise(function (res, rej) {
        rimraf(dirToDelete, function (err) {
            if (err) {
                rej(err);
            }
            else {
                res();
            }
        });
    });
}
exports.deleteDir = deleteDir;
function mkdir(folderName) {
    return new Promise(function (res, rej) {
        mkdirp(folderName, function (err) {
            if (err) {
                rej(err);
            }
            else {
                res();
            }
        });
    });
}
exports.mkdir = mkdir;
function writeFile(fileName, content) {
    return mkdir(path.dirname(fileName))
        .then(function (_) { return util_1.fsAsPromised.writeFile(fileName, content, 'utf8'); });
}
exports.writeFile = writeFile;
