// @stryker-mutator/jest-runner/src/JestTestRunner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var test_runner_1 = require("@stryker-mutator/api/test_runner");
var jestTestAdapters_1 = require("./jestTestAdapters");
var plugin_1 = require("@stryker-mutator/api/plugin");
function jestTestRunnerFactory(injector) {
    return injector
        .provideValue(exports.PROCESS_ENV_TOKEN, process.env)
        .provideValue(jestTestAdapters_1.JEST_VERSION_TOKEN, require('jest/package.json').version)
        .provideFactory(exports.JEST_TEST_ADAPTER_TOKEN, jestTestAdapters_1.jestTestAdapterFactory)
        .injectClass(JestTestRunner);
}
exports.jestTestRunnerFactory = jestTestRunnerFactory;
jestTestRunnerFactory.inject = plugin_1.tokens(plugin_1.commonTokens.injector);
exports.PROCESS_ENV_TOKEN = 'PROCESS_ENV_TOKEN';
exports.JEST_TEST_ADAPTER_TOKEN = 'jestTestAdapter';
var JestTestRunner = /** @class */ (function () {
    function JestTestRunner(log, options, processEnvRef, jestTestAdapter) {
        this.log = log;
        this.processEnvRef = processEnvRef;
        this.jestTestAdapter = jestTestAdapter;
        // Get jest configuration from stryker options and assign it to jestConfig
        this.jestConfig = options.jest.config;
        // Get enableFindRelatedTests from stryker jest options or default to true
        this.enableFindRelatedTests = options.jest.enableFindRelatedTests;
        if (this.enableFindRelatedTests === undefined) {
            this.enableFindRelatedTests = true;
        }
        if (this.enableFindRelatedTests) {
            this.log.debug('Running jest with --findRelatedTests flag. Set jest.enableFindRelatedTests to false to run all tests on every mutant.');
        }
        else {
            this.log.debug('Running jest without --findRelatedTests flag. Set jest.enableFindRelatedTests to true to run only relevant tests on every mutant.');
        }
        // basePath will be used in future releases of Stryker as a way to define the project root
        // Default to process.cwd when basePath is not set for now, should be removed when issue is solved
        // https://github.com/stryker-mutator/stryker/issues/650
        this.jestConfig.rootDir = options.basePath || process.cwd();
        this.log.debug("Project root is " + this.jestConfig.rootDir);
    }
    JestTestRunner.prototype.run = function (options) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var results, errorMessages;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.setNodeEnv();
                        return [4 /*yield*/, this.jestTestAdapter.run(this.jestConfig, process.cwd(), this.enableFindRelatedTests ? options.mutatedFileName : undefined)];
                    case 1:
                        results = (_a.sent()).results;
                        errorMessages = results.testResults.map(function (testSuite) { return testSuite.failureMessage; }).filter(function (errorMessage) { return (errorMessage); });
                        return [2 /*return*/, {
                                errorMessages: errorMessages,
                                status: (results.numRuntimeErrorTestSuites > 0) ? test_runner_1.RunStatus.Error : test_runner_1.RunStatus.Complete,
                                tests: this.processTestResults(results.testResults)
                            }];
                }
            });
        });
    };
    JestTestRunner.prototype.setNodeEnv = function () {
        // Jest CLI will set process.env.NODE_ENV to 'test' when it's null, do the same here
        // https://github.com/facebook/jest/blob/master/packages/jest-cli/bin/jest.js#L12-L14
        if (!this.processEnvRef.NODE_ENV) {
            this.processEnvRef.NODE_ENV = 'test';
        }
    };
    JestTestRunner.prototype.processTestResults = function (suiteResults) {
        var testResults = [];
        for (var _i = 0, suiteResults_1 = suiteResults; _i < suiteResults_1.length; _i++) {
            var suiteResult = suiteResults_1[_i];
            for (var _a = 0, _b = suiteResult.testResults; _a < _b.length; _a++) {
                var testResult = _b[_a];
                testResults.push({
                    failureMessages: testResult.failureMessages,
                    name: testResult.fullName,
                    status: this.determineTestResultStatus(testResult.status),
                    timeSpentMs: testResult.duration ? testResult.duration : 0
                });
            }
        }
        return testResults;
    };
    JestTestRunner.prototype.determineTestResultStatus = function (status) {
        switch (status) {
            case 'passed':
                return test_runner_1.TestStatus.Success;
            case 'pending':
                return test_runner_1.TestStatus.Skipped;
            case 'todo':
                return test_runner_1.TestStatus.Skipped;
            default:
                return test_runner_1.TestStatus.Failed;
        }
    };
    JestTestRunner.inject = plugin_1.tokens(plugin_1.commonTokens.logger, plugin_1.commonTokens.options, exports.PROCESS_ENV_TOKEN, exports.JEST_TEST_ADAPTER_TOKEN);
    return JestTestRunner;
}());
exports.default = JestTestRunner;
