// @stryker-mutator/core/src/reporters/MutationTestReportCalculator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var report_1 = require("@stryker-mutator/api/report");
var plugin_1 = require("@stryker-mutator/api/plugin");
var di_1 = require("../di");
var util_1 = require("@stryker-mutator/util");
var MutationTestReportCalculator = /** @class */ (function () {
    function MutationTestReportCalculator(reporter, options, inputFiles, log) {
        this.reporter = reporter;
        this.options = options;
        this.inputFiles = inputFiles;
        this.log = log;
    }
    MutationTestReportCalculator.prototype.report = function (results) {
        this.reporter.onMutationTestReportReady(this.mutationTestReport(results));
    };
    MutationTestReportCalculator.prototype.mutationTestReport = function (results) {
        return {
            files: this.toFileResults(results),
            schemaVersion: '1.0',
            thresholds: this.options.thresholds
        };
    };
    MutationTestReportCalculator.prototype.toFileResults = function (results) {
        var _this = this;
        var resultDictionary = Object.create(null);
        results.forEach(function (mutantResult) {
            var fileResult = resultDictionary[mutantResult.sourceFilePath];
            if (fileResult) {
                fileResult.mutants.push(_this.toMutantResult(mutantResult));
            }
            else {
                var sourceFile = _this.inputFiles.files.find(function (file) { return file.name === mutantResult.sourceFilePath; });
                if (sourceFile) {
                    resultDictionary[mutantResult.sourceFilePath] = {
                        language: _this.determineLanguage(sourceFile.name),
                        mutants: [_this.toMutantResult(mutantResult)],
                        source: sourceFile.textContent
                    };
                }
                else {
                    _this.log.warn(util_1.normalizeWhitespaces("File \"" + mutantResult.sourceFilePath + "\" not found\n          in input files, but did receive mutant result for it. This shouldn't happen"));
                }
            }
        });
        return resultDictionary;
    };
    MutationTestReportCalculator.prototype.determineLanguage = function (name) {
        var ext = path.extname(name).toLowerCase();
        switch (ext) {
            case '.ts':
            case '.tsx':
                return 'typescript';
            case '.html':
            case '.vue':
                return 'html';
            default:
                return 'javascript';
        }
    };
    MutationTestReportCalculator.prototype.toMutantResult = function (mutantResult) {
        return {
            id: mutantResult.id,
            location: this.toLocation(mutantResult.location),
            mutatorName: mutantResult.mutatorName,
            replacement: mutantResult.replacement,
            status: this.toStatus(mutantResult.status)
        };
    };
    MutationTestReportCalculator.prototype.toLocation = function (location) {
        return {
            end: this.toPosition(location.end),
            start: this.toPosition(location.start)
        };
    };
    MutationTestReportCalculator.prototype.toPosition = function (pos) {
        return {
            column: pos.column + 1,
            line: pos.line + 1
        };
    };
    MutationTestReportCalculator.prototype.toStatus = function (status) {
        switch (status) {
            case report_1.MutantStatus.Killed:
                return "Killed" /* Killed */;
            case report_1.MutantStatus.NoCoverage:
                return "NoCoverage" /* NoCoverage */;
            case report_1.MutantStatus.RuntimeError:
                return "RuntimeError" /* RuntimeError */;
            case report_1.MutantStatus.Survived:
                return "Survived" /* Survived */;
            case report_1.MutantStatus.TimedOut:
                return "Timeout" /* Timeout */;
            case report_1.MutantStatus.TranspileError:
                return "CompileError" /* CompileError */;
            default:
                this.logUnsupportedMutantStatus(status);
                return "RuntimeError" /* RuntimeError */;
        }
    };
    MutationTestReportCalculator.prototype.logUnsupportedMutantStatus = function (status) {
        this.log.warn('Unable to convert "%s" to a MutantStatus', status);
    };
    MutationTestReportCalculator.inject = plugin_1.tokens(di_1.coreTokens.reporter, plugin_1.commonTokens.options, di_1.coreTokens.inputFiles, plugin_1.commonTokens.logger);
    return MutationTestReportCalculator;
}());
exports.MutationTestReportCalculator = MutationTestReportCalculator;
